OpenTabletDriver/Devices/CompositeDeviceHub.cs

```diff
@@ -0,0 +1,143 @@
+                .RegisterServiceProvider(provider);
+        }
+
+        public IEnumerable<IDeviceEndpoint> GetDevices()
+        {
+            return _endpoints;
+        }
+
+        public void ConnectDeviceHub<T>() where T : IDeviceHub
+        {
+            if (_serviceProvider == null)
+                throw new InvalidOperationException("Cannot instantiate device hubs without a service provider");
+
+            ConnectDeviceHub(ActivatorUtilities.CreateInstance<T>(_serviceProvider!));
+        }
+
+        public void ConnectDeviceHub(IDeviceHub deviceHub)
+        {
+            Log.Write(nameof(CompositeDeviceHub), $"Connecting hub: {deviceHub.GetType().Name}", LogLevel.Debug);
+            if (_hubs.Add(deviceHub))
+            {
+                HookDeviceNotification(deviceHub);
+                EnqueueHubChange();
+            }
+            else
+            {
+                Log.Write(nameof(CompositeDeviceHub), $"Connection failed, {deviceHub.GetType().Name} is already connected", LogLevel.Debug);
+            }
+        }
+
+        public void DisconnectDeviceHub<T>() where T : IDeviceHub
+        {
+            foreach (var hub in DeviceHubs.Where(static t => t.GetType().IsAssignableTo(typeof(T))))
+            {
+                DisconnectDeviceHub(hub);
+            }
+        }
+
+        public void DisconnectDeviceHub(IDeviceHub deviceHub)
+        {
+            Log.Write(nameof(CompositeDeviceHub), $"Disconnecting hub: {deviceHub.GetType().Name}", LogLevel.Debug);
+            if (_hubs.Remove(deviceHub))
+            {
+                UnhookDeviceNotification(deviceHub);
+                EnqueueHubChange();
+            }
+            else
+            {
+                Log.Write(nameof(CompositeDeviceHub), $"Disconnection failed, {deviceHub.GetType().Name} is not a connected hub", LogLevel.Debug);
+            }
+        }
+
+        private void HookDeviceNotification(IDeviceHub childHub)
+        {
+            childHub.DevicesChanged += HandleHubDeviceNotification;
+        }
+
+        private void UnhookDeviceNotification(IDeviceHub childHub)
+        {
+            childHub.DevicesChanged -= HandleHubDeviceNotification;
+        }
+
+        private void HandleHubDeviceNotification(object? sender, DevicesChangedEventArgs eventArgs)
+        {
+            Log.Debug(sender?.GetType().Name ?? nameof(CompositeDeviceHub), $"Additions: {eventArgs.Additions.Count()}, Removals: {eventArgs.Removals.Count()}");
+            EnqueueHubChange();
+        }
+
+        private void EnqueueHubChange()
+        {
+            var newCts = new CancellationTokenSource();
+            var oldCts = Interlocked.Exchange(ref _cts, newCts);
+            oldCts?.Cancel();
+
+            Task.Run(() => OnDevicesChanged(newCts.Token), newCts.Token);
+        }
+
+        private async Task OnDevicesChanged(CancellationToken ct)
+        {
+            await Task.Delay(100, ct);
```

Comment:
explain why delay

Reply:
It's for debounce implemented by `EnqueueHubChange()`. Will add comment to clarify that.

Reply:
Resolved